homeassistant/components/binary_sensor/template.py

```diff
@@ -135,4 +147,29 @@ def async_update(self):
                                 "the state is unknown", self._name)
                 return
             _LOGGER.error("Could not render template %s: %s", self._name, ex)
-            self._state = False
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update the state from the template."""
+        state = self._async_render()
+
+        # return if the state don't change or is invalid
+        if state is None or state == self.state:
+            return
+
+        @callback
+        def set_state():
+            """Set state of template binary sensor."""
+            self._state = state
+            self.hass.async_add_job(self.async_update_ha_state())
+
+        # state without delay
+        if (state and not self._delay_on) or \
+                (not state and not self._delay_off):
+            set_state()
```

Comment:
This should not be called from `async_update`. Probably better to rename `async_update` to a different name, what about `async_check_state`